src/lib/tuning/index/index_evaluator.cpp

```diff
@@ -0,0 +1,75 @@
+#include "index_evaluator.hpp"
+
+#include "optimizer/column_statistics.hpp"
+#include "optimizer/table_statistics.hpp"
+#include "resolve_type.hpp"
+#include "storage/index/base_index.hpp"
+#include "storage/storage_manager.hpp"
+#include "storage/table.hpp"
+#include "types.hpp"
+#include "utils/logging.hpp"
+
+namespace opossum {
+
+IndexEvaluator::IndexEvaluator() {}
+
+void IndexEvaluator::_setup() { _saved_work.clear(); }
+
+void IndexEvaluator::_process_access_record(const BaseIndexEvaluator::AccessRecord& record) {
+  auto table_statistics = StorageManager::get().get_table(record.column_ref.table_name)->table_statistics();
+  // ToDo(anyone) adapt for multi column indices...
+  auto predicate_statistics =
+      table_statistics->predicate_statistics(record.column_ref.column_ids[0], record.condition, record.compare_value);
+  auto total_rows = table_statistics->row_count();
+  auto match_rows = predicate_statistics->row_count();
+  auto unscanned_rows = total_rows - match_rows;
+  float saved_work = unscanned_rows * record.query_frequency;
+  std::cout << "saved work for query on " << record.column_ref.table_name << "." << record.column_ref.column_ids[0]
+            << ": " << saved_work << "\n";
+  if (_saved_work.count(record.column_ref) > 0) {
+    _saved_work[record.column_ref] += saved_work;
+  } else {
+    _saved_work[record.column_ref] = saved_work;
+  }
+}
+
+ColumnIndexType IndexEvaluator::_propose_index_type(const IndexChoice& index_evaluation) const {
+  return ColumnIndexType::GroupKey;
+}
+
+float IndexEvaluator::_predict_memory_cost(const IndexChoice& index_evaluation) const {
+  auto table = StorageManager::get().get_table(index_evaluation.column_ref.table_name);
+  // ToDo(anyone) adapt for multi column indices...
+  auto column_statistics = table->table_statistics()->column_statistics().at(index_evaluation.column_ref.column_ids[0]);
+  auto value_count = column_statistics->distinct_count();
+
+  // Sum up column data type widths
+  size_t value_bytes = 0;
+  for (auto column_id : index_evaluation.column_ref.column_ids) {
+    auto data_type = table->column_type(column_id);
+    opossum::resolve_data_type(data_type, [&](auto boost_type) {
+      using ColumnDataType = typename decltype(boost_type)::type;
+      // This assumes that elements are self-contained
+      value_bytes += sizeof(ColumnDataType);
```

Comment:
Why do we need the size of the actual values in the columns? The GroupKey index does rely on dictionary encoding. Did you have other index implementations in mind? Then, never mind.

Reply:
Indeed, the GroupKey index does not need that information. And correct, we wanted to keep this extensible to new index implementations that could work in another way. If you are sure that this is not needed, we could throw it out.